lib/propolis-client/src/instance_spec/devices.rs

```diff
@@ -276,6 +313,14 @@ pub struct DeviceSpec {
     pub network_devices: BTreeMap<SpecKey, NetworkDevice>,
     pub serial_ports: BTreeMap<SpecKey, SerialPort>,
     pub pci_pci_bridges: BTreeMap<SpecKey, PciPciBridge>,
+    #[cfg(feature = "falcon")]
+    pub softnpu_pci_port: Option<SoftNpuPciPort>,
+    #[cfg(feature = "falcon")]
+    pub softnpu_ports: BTreeMap<SpecKey, SoftNpuPort>,
+    #[cfg(feature = "falcon")]
+    pub softnpu_p9: Option<SoftNpuP9>,
+    #[cfg(feature = "falcon")]
+    pub p9fs: Option<P9fs>,
 }
 
 impl DeviceSpec {
```

Comment:
I'm assuming P9 and SoftNpu devices aren't meant to live migrate for now? If they aren't, their `Entity` impls should indicate as much (see my comment over in softnpu.rs). If they are, the `can_migrate_devices_from` routine (which I cannot comment on directly even though it is LITERALLY ON THE NEXT LINE come on Github are you kidding me) should be updated to check these elements for compatibility.

If it existed, the doc comment for `can_migrate_devices_from` would explain that it returns `Ok(())` if `self` and `other` describe specs whose members are similar enough that if you initialized instance 1 from `self` and instance 2 from `other`, you could migrate from one to the other without breaking any guest abstractions (e.g. changing something's PCI path).

Again, my guess is that for now it's best for the components themselves just to say that they can't be migrated, but if you do want them to migrate these checks are important to add.

Reply:
Yeah, we don't need to worry about migration for these devices now. If P9fs support graduates from a Falcon feature to a product feature, we can worry about migration then. For now, Falcon does not migrate things ;)

Reply:
Works for me!